sphynx/python/derive_html.py

```diff
@@ -0,0 +1,51 @@
+'''Run user code.'''
+import io
+import base64
+from matplotlib import pyplot as plt
+import numpy as np
+import pandas as pd
+import os
+import types
+from . import util
+
+op = util.Op()
+if os.environ.get('SPHYNX_CHROOT_PYTHON') == 'yes':
+  op.run_in_chroot()
+
+# Load inputs.
+vs = {}
+es = {}
+graph_attributes = types.SimpleNamespace()
+for fullname in op.inputs.keys():
+  if '.' not in fullname:
+    continue
+  parent, name = fullname.split('.')
+  if parent == 'vs':
+    vs[name] = op.input(fullname)
+  elif parent == 'es':
+    es[name] = op.input(fullname)
+  elif parent == 'graph_attributes':
+    setattr(graph_attributes, name, op.input_scalar(fullname))
+if 'edges-for-es' in op.inputs:
+  edges = op.input('edges-for-es')
+  es['src'] = edges.src
+  es['dst'] = edges.dst
+vs = pd.DataFrame(vs)
+es = pd.DataFrame(es)
+
+# Execute user code.
+try:
+  code = compile(op.params['code'], 'user code', 'exec')
+  exec(code)
+except BaseException:
+  # Hide this file from the traceback.
+  import traceback
+  import sys
+  a, b, c = sys.exc_info()
+  traceback.print_exception(a, b, c.tb_next)
+  sys.exit(1)
+
+f = io.BytesIO()
+plt.savefig(f, format='svg')
+data = base64.b64encode(f.getvalue()).decode('ascii')
+op.output_scalar('sc', f'<img src="data:image/svg+xml;base64, {data}">')
```

Comment:
So the plot is encoded in the HTML.

Because we do this after running the user code, this is not entirely secure. The user could write out their own HTML fragment with a line just like this, and then change `op.output_scalar` to do nothing.

At this point I should clarify what kind of security risk I'm talking about.

If the user can inject arbitrary HTML into the workspace, they can then ask another user to view their workspace. If their injected HTML contains JavaScript, that will be executed under the user credentials of the user viewing the page. ([Cross-site scripting](https://en.wikipedia.org/wiki/Cross-site_scripting))

What we could do to avoid the security issue described in the second paragraph, is to let the Python code just write out the image. Then the Go code would wrap it in an `<img>` tag like on this line. The user would have no way to trick the Go code.

Reply:
> At this point I should clarify what kind of security risk I'm talking about.

The most common method to avoid XSS is to serve content from a different domain. (e.g. githubusercontent.com) We can't exactly do that, but we can still use an `<iframe>`. It turns out iframe has a `sandbox` attribute that makes it behave as if its content is from a different domain. So cool! It also disables scripts and a lot of things by default, which is fine for us for now. But this opens the door to secure interactive Python plots later on.